README.md
# Interologic Python Microservice Template

This repo acts as a template to create restful microservices providing CRUD-based operations in python. 

You can use this repo to get you started with a shell and build upon it. 

## Requirements

* docker >= 2.1.0.0
* pip >= 19.2.3
* pipenv >= 2018.11.26
* python >= 3.7.4

## Running Locally for Development (Quickstart)

### Get Code
```bash
# Clone the repo
git clone git@github.com:interologic/python-microservice-template.git
```

### Setup Virtual Environment
```bash
pipenv shell
```

### Install Dependencies
```bash
pipenv install
```

### Start Development Server
```bash
invoke start
```

## Structure

### Top-level

    .
    ├── .circleci               # circleci configuration
    ├── src                     # source files i.e. routes, models and data access
    ├── tests                   # unit tests
    ├── app.py                  # application entrypoint
    ├── config.py               # configuration class
    ├── Dockerfile              # docker configuration
    ├── Pipfile                 # python dependencies declaration (should be committed to github)
    ├── Pipfile.lock            # lock file to ensure deterministic builds (should also be committed to github)
    ├── task.py                 # defines and invokes various tasks e.g. `invoke start`, `invoke lint`, etc.
    ├── .flake8                 # linting configuration
    ├── .env                    # local configuration settings
    └── README.md

### API

    .
    ├── ...
    ├── api
    │   ├── data                # data access code e.g. database schemas & mapping files 
    │   ├── models              # resource modes that contain business logic, validation & error handling
    │   └── routes              # restful api routes
    └── ...

## Important Routes

### Version
```
/v1/version
```

### Docs
```
/v1/docs
```

### Health
```
/v1/health
```

## Tasks

### Start Development Server
```bash
invoke start
```

### Lint
```bash
invoke lint
```

### Test
```bash
invoke test
```


history.md
1.1.0 / 2019-10-02
==================

  * Pulled in changes from petitions-service to prepare for cip-service

tests/conftest.py
import pytest
from flask import Flask

from src import create_blueprint, create_api
from config import Config


@pytest.fixture
def client():
    """Create and configure a new app instance for each test."""
    app = Flask(__name__)
    app.config.from_object(Config)

    api_blueprint = create_blueprint(app.config)
    create_api(app.config, api_blueprint)
    app.register_blueprint(api_blueprint)

    return app.test_client()


@pytest.fixture
def app():
    app = Flask(__name__)
    return app


tests/routes/test_status.py
import json

from src.routes.status import parse_git
from config import Config

url_base = f'{Config.API_NAME}/{Config.API_URI_VERSION}'


def test_health(client):
    response = client.get(f'{url_base}/health')
    assert json.loads(response.data) == {'status': 'i\'m alive'}


def test_version(client):
    response = client.get(f'{url_base}/version')
    data = json.loads(response.data)
    assert data['git'] == parse_git()
    assert data['name'] == Config.API_NAME
    assert data['version'] == Config.API_SEMANTIC_VERSION


tests/routes/test_template.py
from config import Config

url_base = f'{Config.API_NAME}/{Config.API_URI_VERSION}'


def test_template_great_success(client):
    response = client.get(f'{url_base}/great-success')
    assert response.status_code == 200


def test_template_bad_request(client):
    response = client.get(f'{url_base}/errors/bad-request')
    assert response.status_code == 400


def test_template_conflict(client):
    response = client.get(f'{url_base}/errors/conflict')
    assert response.status_code == 409


def test_template_not_found(client):
    response = client.get(f'{url_base}/errors/not-found')
    assert response.status_code == 404


def test_template_not_implemented(client):
    response = client.get(f'{url_base}/errors/not-implemented')
    assert response.status_code == 501


def test_template_unprocessable_entity(client):
    response = client.get(f'{url_base}/errors/unprocessable-entity')
    assert response.status_code == 422


app.py
from flask import Flask, redirect, url_for, json, current_app
from flask_cors import CORS
from werkzeug.exceptions import HTTPException
from logging.config import dictConfig

from src import create_api, create_blueprint
from extensions import ma

# import config from environment variables
from config import Config

HTTP_STATUS_LOGGING_THRESHOLD = 499

# set logging level
dictConfig({
    'version': 1,
    'root': {
        'level': Config.FLASK_LOGGER_LEVEL
    }
})

# initialize api
app = Flask(__name__)
app.url_map.strict_slashes = False

app.config.from_object(Config)

# Add CORS support
CORS(app)

# initialize marshmallow
ma.init_app(app)

# configure default route
@app.route('/')
@app.route('/' + app.config['API_URI_VERSION'] + '/')
@app.route('/' + app.config['API_NAME'] + '/')
@app.route('/' + app.config['API_NAME'] + '/' + app.config['API_URI_VERSION'] + '/')
def root():
    return redirect(url_for('api.status_version'))


# register api
api_blueprint = create_blueprint(app.config)
api = create_api(app.config, api_blueprint)
app.register_blueprint(api_blueprint)


@app.errorhandler(HTTPException)
def handle_app_exception(ex):
    if ex.code > HTTP_STATUS_LOGGING_THRESHOLD:
        current_app.logger.warning(ex)

    # start with the correct headers and status code from the exception
    response = ex.get_response()

    # replace the body with JSON
    response.content_type = 'application/json'

    data = {
        'code': ex.code,
        'message': ex.name,
        'description': ex.description,
    }

    if hasattr(ex, 'errors') and ex.errors:
        data['errors'] = ex.errors

    response.data = json.dumps(data)

    return response


@api.errorhandler(HTTPException)
def handle_api_exception(ex):
    return ex


if __name__ == '__main__':
    app.run(
        host=app.config['SERVER_HOST'],
        port=app.config['SERVER_PORT'],
        use_reloader=app.config['FLASK_RELOAD']
    )


src/errors.py
from werkzeug.exceptions import HTTPException

BAD_REQUEST_STATUS_CODE = 400
CONFLICT_STATUS_CODE = 409
INTERNAL_SERVER_ERROR_STATUS_CODE = 500
NOT_FOUND_STATUS_CODE = 404
NOT_IMPLEMENTED_STATUS_CODE = 501
UNPROCESSABLE_ENTITY_STATUS_CODE = 422


class BadRequest(HTTPException):
    code = BAD_REQUEST_STATUS_CODE
    message = 'Bad Request'


class Conflict(HTTPException):
    code = CONFLICT_STATUS_CODE
    message = 'Conflict'


class InternalServerError(HTTPException):
    code = INTERNAL_SERVER_ERROR_STATUS_CODE
    message = 'Internal Server Error'


class NotFound(HTTPException):
    code = NOT_FOUND_STATUS_CODE


class NotImplemented(HTTPException):
    code = NOT_IMPLEMENTED_STATUS_CODE
    message = 'Not Implemented'


class UnprocessableEntity(HTTPException):
    code = UNPROCESSABLE_ENTITY_STATUS_CODE
    message = 'Unprocessable Entity'


src/routes/status.py
from datetime import datetime
import subprocess
from flask import current_app
from flask_restplus import Namespace, Resource

start = datetime.now()

router = Namespace(
    'status',
    description='''routes that provide health and version related information'''
)


def uptime():
    duration = (datetime.now() - start).total_seconds()
    result = ''

    days = duration // (24 * 3600)
    duration = duration % (24 * 3600)
    if days > 0:
        result += '%d %s, ' % (days, ('day', 'days')[days == 1])

    hours = duration // 3600
    duration %= 3600
    if days > 0 or hours > 0:
        result += '%d %s, ' % (hours, ('hour', 'hours')[hours == 1])

    minutes = duration // 60
    duration %= 60
    if hours > 0 or minutes > 0:
        result += '%d %s, ' % (minutes, ('minute', 'minutes')[minutes == 1])

    result += '%d %s' % (duration, ('second', 'seconds')[duration == 1])

    return result


def parse_git():
    return {
        'branch': subprocess.check_output(
            ['git', 'rev-parse', '--abbrev-ref', 'HEAD']
        ).decode('ascii').strip(),
        'commit': subprocess.check_output(
            ['git', 'rev-parse', 'HEAD']
        ).decode('ascii').strip()
    }


@router.route('/health')
class Health(Resource):
    def get(self):
        return {'status': 'i\'m alive'}


@router.route('/version')
class Version(Resource):
    def get(self):
        return {
            'git': parse_git(),
            'name': current_app.config['API_NAME'],
            'uptime': uptime(),
            'version': current_app.config['API_SEMANTIC_VERSION']
        }


src/routes/template.py
from flask_restplus import Namespace, Resource
from src.errors import (BadRequest, Conflict, InternalServerError, NotFound, NotImplemented,
                        UnprocessableEntity)


router = Namespace(
    'template',
    description='''routes that show how routes work and should be deleted''',
    validate=False
)


@router.route('/great-success')
class GreatSuccessTest(Resource):
    @router.response(200, 'Success')
    def get(self):
        return {
            'message': 'great success!'
        }, 200


@router.route('/errors/bad-request')
class BadRequestTest(Resource):
    @router.response(400, 'Bad Request')
    def get(self):
        raise BadRequest('Sorry your request is bad :(')


@router.route('/errors/conflict')
class ConflictTest(Resource):
    @router.response(409, 'Conflict')
    def get(self):
        raise Conflict('Your request is conflicted')


@router.route('/errors/internal-server-error')
class InternalServerErrorTest(Resource):
    @router.response(500, 'Internal Server Error')
    def get(self):
        raise InternalServerError('This is real bad')


@router.route('/errors/not-found')
class NotFoundTest(Resource):
    @router.response(404, 'Not Found')
    def get(self):
        raise NotFound('Sorry we couldn\'t find that :(')


@router.route('/errors/not-implemented')
class NotImplementedTest(Resource):
    @router.response(501, 'Not Implemented')
    def get(self):
        raise NotImplemented('This route is not implemented yet')


@router.route('/errors/unprocessable-entity')
class UnprocessableEntityTest(Resource):
    @router.response(422, 'Unprocessable Entity')
    def get(self):
        ex = UnprocessableEntity('Nope can\'t do it')
        ex.errors = [
            {
                'code': 5432,
                'field': 'first_name',
                'message': 'You must provide your first name'
            }
        ]
        raise ex
